src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,154 @@
+use std::marker::PhantomData;
+
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+    },
+    protocol::{context::ProtocolContext, IPAProtocolStep, RecordId, SortStep, Step},
+    secret_sharing::Replicated,
+};
+
+use futures::future::join_all;
+
+#[derive(Debug)]
+pub struct BitPermutations<'a, G, M, S: Step + SpaceIndex> {
+    pub(crate) context: &'a ProtocolContext<'a, G, S>,
+    pub(crate) phantom: PhantomData<M>,
```

Comment:
I think we can make it a bit more ergonomic to use if we invert the state and keep the input here instead of `context`.

```rust
#[derive(Debug)]
pub struct BitPermutations<'a, F> {
    input: &'a [Replicated<F>]
}
```

I have a sketch ready if you want to see how it would look like https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4359f076c0dfccc632e67e4ccf422166

let me know what you think

Reply:
Looks good! Lot of boilerplate code got removed e.g. PhantomData and interface to call BitPermutations looks much cleaner!

Reply:
However, this pattern is opposite of what is done in securemul i.e. it passes inputs to execute vs this passes inputs to struct declaration. Should we have similar approach in both declarations?

Reply:
I believe it is ok for them to be different - secure mul is just a very important primitive to be used everywhere, so we may implement it differently to keep it ergonomic to use